### PR TITLE
Support loading via NDJSON in CartoLayer

### PR DESCRIPTION
**Pull request by felixpalmer**
Implementation of https://github.com/visgl/deck.gl/issues/6201

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -2038,6 +2038,15 @@
   dependencies:
     "@loaders.gl/loader-utils" "3.0.8"
 
+"@loaders.gl/loader-utils@3.0.12":
```
Comment (Pessimistress) [inline yarn.lock:2041]: You have multiple versions of loaders installed

**Review by alasarr — CHANGES_REQUESTED** [review]
Multiple version of loaders.gl should be fixed, the rest looks good to me

File: modules/carto/src/api/maps-v3-client.js [hunk 2]
```diff
@@ -168,5 +192,10 @@ export async function getData({type, source, connection, credentials, geoColumn,
 
   const {accessToken} = localCreds;
 
-  return await request({url, format: mapFormat, accessToken});
+  const data = await request({url, format: mapFormat, accessToken});
+  if (mapFormat === FORMATS.CSV) {
+    return csvToGeoJson(data, {geoColumn});
```
Comment (alasarr) [inline modules/carto/src/api/maps-v3-client.js:197]: It's a little bit strange that getData returns the data in geoJson format when you asked for CSV.  Have in mind that this function is exposed publicly: https://deck.gl/docs/api-reference/carto/overview#support-for-other-deckgl-layers.

File: yarn.lock (hunk 1 above)
Comment (felixpalmer) [reply yarn.lock]: I'm at a loss on how to get yarn to do what I want here. I've updated `package.json` to pick a specific version of wkt (`"@loaders.gl/wkt": "3.0.8"`) and then there is only one version of loaders. But when I set it to `^3.0.8` like the other packages, yarn insists on pulling in 3.0.12 in some places.

File: modules/carto/src/api/maps-v3-client.js (hunk 2 above)
Comment (felixpalmer) [reply modules/carto/src/api/maps-v3-client.js]: I have done this now, but I changed the signature of `getData` to also return the format. Is this OK? Without also explictly returning the format I don't how we can cleanly do the conversion outside the `getData` function as the format is determined by what the map instantiation request returns

File: yarn.lock (hunk 1 above)
Comment (felixpalmer) [reply yarn.lock]: This is now no longer relevant. Please ignore above comment

**Comment by alasarr** [comment]
@felixpalmer as we've talked before, let's remove the CSV format and implement only the NDJSON stream in CartoLayer 💪 . It looks promising

**Comment by felixpalmer** [comment]
@alasarr CSV loading is removed. As discussed, we need to wait for loaders.gl 3.1 to be published (and bumped in deck.gl) in order to merge this PR - as `NDJSONLoader` is only available in the alpha package.

File: modules/carto/src/api/maps-v3-client.js [hunk 3]
```diff
@@ -123,9 +123,7 @@ function getUrlFromMetadata(metadata, format) {
   return null;
 }
 
-export async function getData({type, source, connection, credentials, geoColumn, columns, format}) {
-  const localCreds = {...getDefaultCredentials(), ...credentials};
-
+function checkGetDataParameters({type, source, connection, localCreds}) {
```
Comment (alasarr) [inline modules/carto/src/api/maps-v3-client.js:126]: 👍

File: modules/carto/src/api/maps-v3-client.js [hunk 4]
```diff
@@ -168,5 +167,6 @@ export async function getData({type, source, connection, credentials, geoColumn,
 
   const {accessToken} = localCreds;
 
-  return await request({url, format: mapFormat, accessToken});
+  const data = await request({url, format: mapFormat, accessToken});
+  return {data, format: mapFormat};
```
Comment (alasarr) [inline modules/carto/src/api/maps-v3-client.js:171]: You need to modify this example: https://deck.gl/docs/api-reference/carto/overview#support-for-other-deckgl-layers

File: modules/carto/src/layers/carto-layer.js [hunk 5]
```diff
@@ -100,14 +103,31 @@ export default class CartoLayer extends CompositeLayer {
...
+
+        if (result.format === FORMATS.NDJSON) {
+          const batches = await parseInBatches(result.data, NDJSONLoader, {
+            batchSize: 5000
+          });
+          const wait = r => setTimeout(r, 0);
+          const getDataAsync = async function* getDataAsync() {
+            for await (const batch of batches) {
+              // To avoid UI locking up, only yield once per frame
+              await new Promise(wait);
```
Comment (alasarr) [inline modules/carto/src/layers/carto-layer.js:123]: Why do we need it?

**Review by alasarr — CHANGES_REQUESTED** [review]

File: modules/carto/src/api/maps-v3-client.js (hunk 4 above)
Comment (felixpalmer) [reply modules/carto/src/api/maps-v3-client.js:171]: Done: https://github.com/visgl/deck.gl/pull/6202/commits/f76ee9d8b3f3530dd80afe9933f407e7296d0861

File: modules/carto/src/layers/carto-layer.js (hunk 5 above)
Comment (felixpalmer) [reply modules/carto/src/layers/carto-layer.js:123]: Well, it is optional. I just felt that without it the UI feels a bit sluggish when loading a large dataset. But it is of course a tradeoff, if it is removed the full dataset is displayed quicker. What do you think, which should we optimize for?

File: docs/api-reference/carto/overview.md [hunk 6]
```diff
@@ -151,7 +152,12 @@ The CARTO submodule includes the CartoLayer that simplifies the interaction with
     });
     ```
 
-    The formats available are JSON, GEOJSON, TILEJSON, and NDJSON. [NDJSON](http://ndjson.org/) (Newline Delimited JSON) allows to handle incremental data loading https://deck.gl/docs/developer-guide/performance#handle-incremental-data-loading.
+    The formats available are JSON, GEOJSON, TILEJSON, and NDJSON. [NDJSON](http://ndjson.org/) (Newline Delimited JSON) allows to handle incremental data loading https://deck.gl/docs/developer-guide/performance#handle-incremental-data-loading. 
+
+* If the format is not explicitly specified, `getData` will pick a format automatically, depending on what is available from the CARTO API. The `getData` function returns a `Object` with the following properties:
```
Comment (borja-munoz) [inline docs/api-reference/carto/overview.md:157]: If possible, I'd add another level of indentation for clarity. The first indentation level should describe just the two options available (using API v3 or not)

File: docs/api-reference/carto/overview.md [hunk 7]
```diff
@@ -133,14 +133,15 @@ The CARTO submodule includes the CartoLayer that simplifies the interaction with
...
     import { H3HexagonLayer } from '@deck.gl/geo-layers/';
 
-    const data =  await getData({
+    const result = await getData({
       type: MAP_TYPES.QUERY,
       source: `SELECT bqcarto.h3.ST_ASH3(internal_point_geom, 4) as h3, count(*) as count
                   FROM bigquery-public-data.geo_us_census_places.us_national_places 
                 GROUP BY h3`,
       connection: 'connection_name',
       format: 'json'
```
Comment (borja-munoz) [inline docs/api-reference/carto/overview.md:142]: I'd update this line to use the corresponding constant from the FORMATS enumeration

**Comment by borja-munoz** [comment]
The change in the getData returned object is a breaking change, so we should also update the "Upgrade Guide".

**Review by alasarr — APPROVED** [review]

**Comment by felixpalmer** [comment]
As discussed, we need to wait for loaders.gl 3.1 to be published (and bumped in deck.gl) in order to merge this PR - as NDJSONLoader is only available in the alpha package.

**Pull request closed without merging**